File: controllers/admin_controller.go
package controllers

import (
	"errors"
	"net/http"

	"backend-sipeka/config"
	"backend-sipeka/models"

	"github.com/gin-gonic/gin"
	"golang.org/x/crypto/bcrypt"
	"gorm.io/gorm"
)

// CreateUserInternalInput mendefinisikan payload untuk pembuatan user oleh Admin
type CreateUserInternalInput struct {
	NisnNip  string `json:"nisn_nip" binding:"required"`
	Name     string `json:"name" binding:"required"`
	Password string `json:"password" binding:"required"`
	Role     string `json:"role" binding:"required"`
}

// CreateUserInternal membuat akun internal baru (khusus kasir atau admin) oleh Admin
func CreateUserInternal(c *gin.Context) {
	var input CreateUserInternalInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, models.Response{
			Status:  "error",
			Message: "Format data request tidak valid: nisn_nip, name, password, dan role wajib diisi",
		})
		return
	}

	// Validasi role internal yang diizinkan (kasir atau admin)
	if input.Role != "kasir" && input.Role != "admin" {
		c.JSON(http.StatusBadRequest, models.Response{
			Status:  "error",
			Message: "Pembuatan akun internal hanya diperbolehkan untuk role 'kasir' atau 'admin'",
		})
		return
	}

	// Cek apakah NISN/NIP sudah terdaftar
	var existingUser models.User
	if err := config.DB.Where("nisn_nip = ?", input.NisnNip).First(&existingUser).Error; err == nil {
		c.JSON(http.StatusConflict, models.Response{
			Status:  "error",
			Message: "NISN/NIP sudah terdaftar dalam sistem",
		})
		return
	}

	// Hash password menggunakan bcrypt
	hashedPass, err := bcrypt.GenerateFromPassword([]byte(input.Password), bcrypt.DefaultCost)
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.Response{
			Status:  "error",
			Message: "Gagal mengenkripsi kata sandi",
		})
		return
	}

	// Set is_active = true secara default untuk akun internal yang dibuat admin
	newUser := models.User{
		NisnNip:      input.NisnNip,
		Name:         input.Name,
		PasswordHash: string(hashedPass),
		Role:         input.Role,
		IsActive:     true,
	}

	if err := config.DB.Create(&newUser).Error; err != nil {
		c.JSON(http.StatusInternalServerError, models.Response{
			Status:  "error",
			Message: "Gagal menyimpan akun pengguna internal ke database",
		})
		return
	}

	c.JSON(http.StatusCreated, models.Response{
		Status:  "success",
		Message: "Akun pengguna internal berhasil dibuat",
		Data: UserResponseData{
			ID:       newUser.ID,
			NisnNip:  newUser.NisnNip,
			Name:     newUser.Name,
			Role:     newUser.Role,
			IsActive: newUser.IsActive,
		},
	})
}

// ApproveUser menyetujui dan mengaktifkan akun pengguna (terutama penitip) berdasarkan ID
func ApproveUser(c *gin.Context) {
	id := c.Param("id")
	if id == "" {
		c.JSON(http.StatusBadRequest, models.Response{
			Status:  "error",
			Message: "Parameter ID pengguna tidak boleh kosong",
		})
		return
	}

	var user models.User
	err := config.DB.Where("id = ?", id).First(&user).Error
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			c.JSON(http.StatusNotFound, models.Response{
				Status:  "error",
				Message: "Pengguna dengan ID tersebut tidak ditemukan",
			})
			return
		}
		c.JSON(http.StatusInternalServerError, models.Response{
			Status:  "error",
			Message: "Terjadi kesalahan internal saat mencari data pengguna",
		})
		return
	}

	// Update is_active = true
	if err := config.DB.Model(&user).Update("is_active", true).Error; err != nil {
		c.JSON(http.StatusInternalServerError, models.Response{
			Status:  "error",
			Message: "Gagal mengaktifkan akun pengguna",
		})
		return
	}
	user.IsActive = true

	c.JSON(http.StatusOK, models.Response{
		Status:  "success",
		Message: "Akun pengguna berhasil disetujui dan diaktifkan",
		Data: UserResponseData{
			ID:       user.ID,
			NisnNip:  user.NisnNip,
			Name:     user.Name,
			Role:     user.Role,
			IsActive: user.IsActive,
		},
	})
}

// GetUsers mengambil daftar seluruh pengguna dengan filter opsional (role, status is_active)
func GetUsers(c *gin.Context) {
	var users []models.User
	query := config.DB.Model(&models.User{})

	if role := c.Query("role"); role != "" {
		query = query.Where("role = ?", role)
	}
	if active := c.Query("is_active"); active != "" {
		if active == "true" || active == "1" {
			query = query.Where("is_active = ?", true)
		} else if active == "false" || active == "0" {
			query = query.Where("is_active = ?", false)
		}
	}

	if err := query.Order("created_at DESC").Find(&users).Error; err != nil {
		c.JSON(http.StatusInternalServerError, models.Response{
			Status:  "error",
			Message: "Gagal mengambil daftar pengguna",
		})
		return
	}

	var result []UserResponseData
	for _, u := range users {
		result = append(result, UserResponseData{
			ID:       u.ID,
			NisnNip:  u.NisnNip,
			Name:     u.Name,
			Role:     u.Role,
			IsActive: u.IsActive,
		})
	}

	c.JSON(http.StatusOK, models.Response{
		Status:  "success",
		Message: "Daftar pengguna berhasil diambil",
		Data:    result,
	})
}

File: controllers/adminController.js
const bcrypt = require('bcryptjs');
const { User } = require('../models');

/**
 * Menyetujui pendaftaran pengguna (misal akun penitip) sehingga is_active menjadi true
 */
const approveUser = async (req, res) => {
  try {
    const userId = req.params.id || req.body.id || req.body.user_id;

    if (!userId) {
      return res.status(400).json({
        status: 'error',
        message: 'ID pengguna wajib disertakan (via parameter URL atau body).',
        data: null
      });
    }

    const user = await User.findByPk(userId);
    if (!user) {
      return res.status(404).json({
        status: 'error',
        message: `Pengguna dengan ID '${userId}' tidak ditemukan.`,
        data: null
      });
    }

    if (user.is_active) {
      return res.status(200).json({
        status: 'success',
        message: 'Pengguna ini sudah dalam status aktif.',
        data: {
          id: user.id,
          nisn_nip: user.nisn_nip,
          name: user.name,
          role: user.role,
          is_active: user.is_active
        }
      });
    }

    user.is_active = true;
    await user.save();

    return res.status(200).json({
      status: 'success',
      message: `Akun pengguna '${user.name}' (${user.role}) berhasil disetujui dan diaktifkan.`,
      data: {
        id: user.id,
        nisn_nip: user.nisn_nip,
        name: user.name,
        role: user.role,
        is_active: user.is_active
      }
    });
  } catch (error) {
    return res.status(500).json({
      status: 'error',
      message: 'Terjadi kesalahan saat menyetujui pengguna: ' + error.message,
      data: null
    });
  }
};

/**
 * Membuat akun pengguna internal (kasir atau admin) langsung aktif (is_active = true)
 */
const createInternalUser = async (req, res) => {
  try {
    const { nisn_nip, name, password, role } = req.body;

    if (!nisn_nip || !name || !password || !role) {
      return res.status(400).json({
        status: 'error',
        message: 'Field nisn_nip, name, password, dan role wajib diisi.',
        data: null
      });
    }

    const normalizedRole = role.toLowerCase();

    if (normalizedRole !== 'kasir' && normalizedRole !== 'admin') {
      return res.status(400).json({
        status: 'error',
        message: "Pembuatan akun internal hanya diperuntukkan bagi role 'kasir' atau 'admin'.",
        data: null
      });
    }

    // Periksa apakah NISN/NIP sudah terdaftar
    const existingUser = await User.findOne({ where: { nisn_nip } });
    if (existingUser) {
      return res.status(400).json({
        status: 'error',
        message: `Pengguna dengan NISN/NIP '${nisn_nip}' sudah terdaftar.`,
        data: null
      });
    }

    // Hash kata sandi
    const salt = await bcrypt.genSalt(10);
    const password_hash = await bcrypt.hash(password, salt);

    // Akun internal langsung aktif (is_active = true)
    const newUser = await User.create({
      nisn_nip,
      name,
      password_hash,
      role: normalizedRole,
      is_active: true
    });

    return res.status(201).json({
      status: 'success',
      message: `Akun internal '${newUser.name}' dengan role '${newUser.role}' berhasil dibuat.`,
      data: {
        id: newUser.id,
        nisn_nip: newUser.nisn_nip,
        name: newUser.name,
        role: newUser.role,
        is_active: newUser.is_active
      }
    });
  } catch (error) {
    return res.status(500).json({
      status: 'error',
      message: 'Terjadi kesalahan saat membuat pengguna internal: ' + error.message,
      data: null
    });
  }
};

/**
 * Mendapatkan seluruh daftar pengguna (dengan opsi filter)
 */
const getAllUsers = async (req, res) => {
  try {
    const { role, is_active } = req.query;
    const whereClause = {};

    if (role) {
      whereClause.role = role.toLowerCase();
    }

    if (is_active !== undefined) {
      whereClause.is_active = is_active === 'true' || is_active === '1';
    }

    const users = await User.findAll({
      where: whereClause,
      attributes: ['id', 'nisn_nip', 'name', 'role', 'is_active', 'createdAt', 'updatedAt'],
      order: [['createdAt', 'DESC']]
    });

    return res.status(200).json({
      status: 'success',
      message: 'Daftar pengguna berhasil diambil.',
      data: users
    });
  } catch (error) {
    return res.status(500).json({
      status: 'error',
      message: 'Gagal mengambil data pengguna: ' + error.message,
      data: null
    });
  }
};

module.exports = {
  approveUser,
  createInternalUser,
  getAllUsers
};
